wallet/udb/testdata/v8.db.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/decred/dcrd/dcrec/secp256k1"
+	"github.com/decred/dcrd/dcrutil"
+	"github.com/decred/dcrd/hdkeychain"
+	"github.com/decred/dcrd/txscript"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrwallet/wallet/udb"
+	"github.com/decred/dcrwallet/walletdb"
+	_ "github.com/decred/dcrwallet/walletdb/bdb"
+	"github.com/decred/dcrwallet/walletseed"
+)
+
+const dbname = "v8.db"
+
+var (
+	epoch    time.Time
+	pubPass  = []byte("public")
+	privPass = []byte("private")
+	privKey  = []byte{31: 1}
+)
+
+func main() {
+	err := setup()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "setup: %v\n", err)
+		os.Exit(1)
+	}
+	err = compress()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "compress: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+// addSStxChange adds a new output with the given amount and address
+func addSStxChange(msgtx *wire.MsgTx, change dcrutil.Amount,
+	changeAddr dcrutil.Address) error {
+	pkScript, err := txscript.PayToSStxChange(changeAddr)
+	if err != nil {
+		return fmt.Errorf("cannot create txout script: %s", err)
+	}
+	msgtx.AddTxOut(wire.NewTxOut(int64(change), pkScript))
+	return nil
+}
+
+func setup() error {
+	var chainParams = &chaincfg.TestNet2Params
+	os.Remove(dbname)
+	db, err := walletdb.Create("bdb", dbname)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+	seed, err := walletseed.GenerateRandomSeed(hdkeychain.RecommendedSeedLen)
+	if err != nil {
+		return err
+	}
+	err = udb.Initialize(db, chainParams, seed, pubPass, privPass)
+	if err != nil {
+		return err
+	}
+
+	amgr, txmgr, _, err := udb.Open(db, chainParams, pubPass)
+	if err != nil {
+		return err
+	}
+
+	return walletdb.Update(db, func(dbtx walletdb.ReadWriteTx) error {
+		amgrns := dbtx.ReadWriteBucket([]byte("waddrmgr"))
+		txmgrns := dbtx.ReadWriteBucket([]byte("wtxmgr"))
+		err := amgr.Unlock(amgrns, privPass)
+		if err != nil {
+			return err
+		}
+
+		privKey, _ := secp256k1.PrivKeyFromBytes(privKey)
+		wif, err := dcrutil.NewWIF(privKey, chainParams, chainec.ECTypeSecp256k1)
+		if err != nil {
+			return err
+		}
+		maddr, err := amgr.ImportPrivateKey(amgrns, wif)
+		if err != nil {
+			return err
+		}
+		addr := maddr.Address()
+
+		// Add a block
+		prevBlock := chainParams.GenesisHash
+		buf := bytes.Buffer{}
+		err = (&wire.BlockHeader{
+			Version:      1,
+			PrevBlock:    *prevBlock,
+			StakeVersion: 1,
+			VoteBits:     1,
+			Height:       uint32(1),
+		}).Serialize(&buf)
+		if err != nil {
+			return err
+		}
+
+		headerData := udb.BlockHeaderData{
+			BlockHash: chainhash.Hash{31: byte(1)},
+		}
+		copy(headerData.SerializedHeader[:], buf.Bytes())
+		err = txmgr.ExtendMainChain(txmgrns, &headerData)
+		if err != nil {
+			return err
+		}
+
+		blockMeta := &udb.BlockMeta{
+			Block: udb.Block{
+				Hash:   headerData.BlockHash,
+				Height: int32(1),
+			},
+			Time: epoch,
+		}
+
+		// Add 6 sstxchange outputs to (3 mined, 3 unmined)
+		for count := 1; count < 7; count++ {
+			msgTx := &wire.MsgTx{}
```

Comment:
use wire.NewMsgTx here.  default-constructed transactions have introduced bugs before since they don't explicitly specify the tx version, and when needed, that API will be updated to take the tx version as a parameter.